go/private/actions/link.bzl

```diff
@@ -136,16 +137,27 @@ def emit_link(
     extldflags.extend(cgo_rpaths)
 
     # Process x_defs, and record whether stamping is used.
-    stamp_x_defs = False
+    stamp_x_defs_volatile = False
+    stamp_x_defs_stable = False
     for k, v in archive.x_defs.items():
-        if go.stamp and v.find("{") != -1 and v.find("}") != -1:
-            stamp_x_defs = True
         builder_args.add("-X", "%s=%s" % (k, v))
+        if go.stamp:
+            stable_vars_count = (count_group_matches(v, "{STABLE_", "}") +
+                                 v.count("{BUILD_EMBED_LABEL}") +
+                                 v.count("{BUILD_USER}") +
+                                 v.count("{BUILD_HOST}"))
+            if stable_vars_count > 0:
+                stamp_x_defs_stable = True
+            if count_group_matches(v, "{", "}") != stable_vars_count:
+                stamp_x_defs_volatile = True
```

Comment:
I was a bit confused why do we need `count_group_matches` at all and this seems to be overly complicated. But then I think I got the reason:

`x_defs` stamping comes as a key-value pair, something like

```
go_binary(
    name = "cmd",
    srcs = ["main.go"],
    deps = ["//version"],
    x_defs = {"example.com/repo/version.Version": "{STABLE_GIT_COMMIT}"},
)
```

and although most documentation have put value as a variable surrounded with braces `{}`, in truth we do allow multiple variables in 1 value, such as 

```
    x_defs = {"example.com/repo/version.Version": "{STABLE_GIT_COMMIT}_{MY_VOLATILE_VAL}"},
```
is still valid 🤔

This is ultimately following the behavior of the custom go linker wrapper https://sourcegraph.com/github.com/bazelbuild/rules_go@e9a7054ff11a520e3b8aceb76a3ba44bb8da4c94/-/blob/go/tools/builders/link.go#L124-140 which uses the regex `{.+?}`